test/unit/eigensolver/mc/test_backtransformation.cpp

```diff
@@ -0,0 +1,172 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2021, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#include "dlaf/eigensolver/backtransformation.h"
+
+#include <functional>
+#include <sstream>
+#include <tuple>
+#include "gtest/gtest.h"
+#include "dlaf/common/index2d.h"
+#include "dlaf/matrix/copy.h"
+#include "dlaf/matrix/matrix.h"
+#include "dlaf/matrix/matrix_base.h"
+#include "dlaf/types.h"
+#include "dlaf/util_matrix.h"
+#include "dlaf_test/matrix/matrix_local.h"
+#include "dlaf_test/matrix/util_matrix.h"
+#include "dlaf_test/matrix/util_matrix_blas.h"
+#include "dlaf_test/matrix/util_matrix_local.h"
+#include "dlaf_test/util_types.h"
+
+using namespace dlaf;
+using namespace dlaf::common;
+using namespace dlaf::matrix;
+using namespace dlaf::matrix::internal;
+using namespace dlaf::matrix::test;
+using namespace dlaf::test;
+using dlaf::matrix::test::MatrixLocal;
+using namespace testing;
+
+template <typename Type>
+class BackTransformationEigenSolverTestMC : public ::testing::Test {};
+TYPED_TEST_SUITE(BackTransformationEigenSolverTestMC, MatrixElementTypes);
+
+GlobalElementSize globalTestSize(const LocalElementSize& size) {
+  return {size.rows(), size.cols()};
+}
+
+const std::vector<std::tuple<SizeType, SizeType, SizeType, SizeType>> sizes =
+    // m, n, mb, nb
+    {{3, 0, 1, 1},   {0, 5, 2, 3},  // m, n = 0
+     {2, 2, 3, 3},   {3, 4, 6, 7},  // m < mb
+     {3, 3, 1, 1},   {4, 4, 2, 2}, {6, 3, 3, 3}, {12, 2, 4, 4}, {12, 24, 3, 3},
+     {24, 36, 6, 6}, {5, 8, 3, 2}, {4, 6, 2, 3}, {5, 5, 2, 3},  {8, 27, 3, 4},
+     {15, 34, 4, 6}, {7, 6, 2, 3}, {8, 5, 2, 3}, {7, 5, 2, 3}};
+
+template <class T>
+MatrixLocal<T> makeLocal(const Matrix<const T, Device::CPU>& matrix) {
+  return {matrix.size(), matrix.distribution().blockSize()};
+}
+
+template <class T>
+void set_zero(Matrix<T, Device::CPU>& mat) {
+  dlaf::matrix::util::set(mat, [](auto&&) { return static_cast<T>(0.0); });
+}
+
+template <class T>
+void getTau(T& tau, T dotprod, BaseType<T> /*tau_i*/) {
+  tau = static_cast<T>(2.0) / dotprod;
+}
+
+template <class T>
+void getTau(std::complex<T>& tau, T dotprod, BaseType<T> tau_i) {
+  tau = {(static_cast<T>(1.0) + sqrt(static_cast<T>(1.0) - dotprod * tau_i * tau_i)) / dotprod, tau_i};
+}
+
+template <class T>
+void testBacktransformationEigenv(SizeType m, SizeType n, SizeType mb, SizeType nb) {
+  LocalElementSize sizeC(m, n);
+  TileElementSize blockSizeC(mb, nb);
+  Matrix<T, Device::CPU> mat_c(sizeC, blockSizeC);
+  dlaf::matrix::util::set_random(mat_c);
+
+  LocalElementSize sizeV(m, m);
+  TileElementSize blockSizeV(mb, mb);
+  Matrix<T, Device::CPU> mat_v(sizeV, blockSizeV);
+  dlaf::matrix::util::set_random(mat_v);
+
+  SizeType tottaus;
+  if (m < mb || m == 0 || n == 0)
+    tottaus = 0;
+  else
+    tottaus = m - mb;
+
+  if (tottaus > 0) {

```

Comment:
If no HHR has to be applied it should be checked that C doesn't change. (Suggested: apply the test even if tottaus = 0).

```suggestion
  if (m == 0 || n == 0)
    return;

  const SizeType tottaus = std::max(0, m - mb);
```

Moreover as suggested above, consider renaming tottaus.